File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/ingredients_analysis/IngredientsWithTagDialogFragment.java
package openfoodfacts.github.scrachx.openfood.features.product.view.ingredients_analysis;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.text.Html;
import android.text.Spanned;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatButton;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.SwitchCompat;
import androidx.browser.customtabs.CustomTabsIntent;
import androidx.fragment.app.DialogFragment;
import androidx.preference.PreferenceManager;

import org.apache.commons.lang.StringUtils;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

import openfoodfacts.github.scrachx.openfood.R;
import openfoodfacts.github.scrachx.openfood.customtabs.CustomTabActivityHelper;
import openfoodfacts.github.scrachx.openfood.features.product.view.ProductViewActivity;
import openfoodfacts.github.scrachx.openfood.features.scan.ContinuousScanActivity;
import openfoodfacts.github.scrachx.openfood.models.Product;
import openfoodfacts.github.scrachx.openfood.models.entities.analysistagconfig.AnalysisTagConfig;
import openfoodfacts.github.scrachx.openfood.utils.Utils;

public class IngredientsWithTagDialogFragment extends DialogFragment {
    private static final String TAG_KEY = "tag";
    private static final String TYPE_KEY = "type";
    private static final String TYPE_NAME_KEY = "type_name";
    private static final String ICON_URL_KEY = "icon_url";
    private static final String COLOR_KEY = "color";
    private static final String NAME_KEY = "name";
    private static final String INGREDIENTS_IMAGE_URL_KEY = "ingredients_image_url";
    private static final String INGREDIENTS_KEY = "ingredients";
    private static final String MISSING_INGREDIENTS_KEY = "missing_ingredients";
    private static final String PHOTOS_TO_BE_VALIDATED_KEY = "photos_to_be_validated";
    private static final String AMBIGUOUS_INGREDIENT_KEY = "ambiguous_ingredient";
    private DialogInterface.OnDismissListener onDismissListener;

    public static IngredientsWithTagDialogFragment newInstance(Product product, AnalysisTagConfig config) {
        IngredientsWithTagDialogFragment frag = new IngredientsWithTagDialogFragment();
        Bundle args = new Bundle();
        args.putString(TAG_KEY, config.getAnalysisTag());
        args.putString(TYPE_KEY, config.getType());
        args.putString(TYPE_NAME_KEY, config.getTypeName());
        args.putString(ICON_URL_KEY, config.getIconUrl());
        args.putString(COLOR_KEY, config.getColor());
        args.putString(NAME_KEY, config.getName().getName());
        args.putString(INGREDIENTS_IMAGE_URL_KEY, product.getImageIngredientsUrl());

        if (product.getIngredients() == null || product.getIngredients().isEmpty()) {
            final List<String> statesTags = product.getStatesTags();
            boolean ingredientsToBeCompleted = false;
            boolean photosToBeValidated = false;

            for (String stateTag : statesTags) {
                if (stateTag.equals("en:ingredients-to-be-completed")) {
                    ingredientsToBeCompleted = true;
                } else if (stateTag.equals("en:photos-to-be-validated")) {
                    photosToBeValidated = true;
                }
            }

            if (ingredientsToBeCompleted && photosToBeValidated) {
                args.putBoolean(PHOTOS_TO_BE_VALIDATED_KEY, true);
            } else {
                args.putBoolean(MISSING_INGREDIENTS_KEY, true);
            }
        } else {
            String showIngredients = config.getName().getShowIngredients();
            if (showIngredients != null) {
                args.putSerializable(INGREDIENTS_KEY, getMatchingIngredientsText(product, showIngredients.split(":")));
            }
            List<String> ambiguousIngredient = product.getIngredients().stream()
                .filter(ingredientList -> ingredientList.containsKey(config.getType()) && ingredientList.containsValue("maybe"))
                .filter(ingredient -> ingredient.containsKey("text")).map(ingredient -> ingredient.get("text"))
                .collect(Collectors.toList());
            if (!ambiguousIngredient.isEmpty()) {
                args.putString(AMBIGUOUS_INGREDIENT_KEY, StringUtils.join(ambiguousIngredient, ","));
            }
        }
        frag.setArguments(args);
        return frag;
    }

    private static String getMatchingIngredientsText(Product product, String[] showIngredients) {
        ArrayList<String> matchingIngredients = new ArrayList<>();
        List<LinkedHashMap<String, String>> ingredients = product.getIngredients();
        for (LinkedHashMap<String, String> ingredient :
            ingredients) {
            if (showIngredients[1].equals(ingredient.get(showIngredients[0]))) {
                final String text = ingredient.get("text");
                if (text != null) {
                    matchingIngredients.add(text.toLowerCase().replace("_", ""));
                }
            }
        }

        if (matchingIngredients.isEmpty()) {
            return null;
        }

        StringBuilder text = new StringBuilder(128);
        text.append(" <b>");
        text.append(matchingIngredients.get(0));
        for (int i = 1; i < matchingIngredients.size(); ++i) {
            text.append(", ");
            text.append(matchingIngredients.get(i));
        }
        text.append("</b>");

        return text.toString();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

        final View rootView = inflater.inflate(R.layout.ingredients_with_tag, container);
        requireDialog().getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        requireDialog().getWindow().setGravity(Gravity.CENTER);

        if (getActivity() == null) {
            return rootView;
        }
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(requireActivity());

        final Bundle arguments = requireArguments();
        String tag = arguments.getString(TAG_KEY);
        String type = arguments.getString(TYPE_KEY);
        String typeName = arguments.getString(TYPE_NAME_KEY);
        String iconUrl = arguments.getString(ICON_URL_KEY);
        String color = arguments.getString(COLOR_KEY);
        String name = arguments.getString(NAME_KEY);
        String ingredients = arguments.getString(INGREDIENTS_KEY);
        String ambiguousIngredient = arguments.getString(AMBIGUOUS_INGREDIENT_KEY);

        AppCompatImageView icon = rootView.findViewById(R.id.icon);
        Utils.picassoBuilder(getActivity())
            .load(iconUrl)
            .into(icon);
        Drawable background = getResources().getDrawable(R.drawable.rounded_button);
        background.setColorFilter(Color.parseColor(color), android.graphics.PorterDuff.Mode.SRC_IN);
        rootView.findViewById(R.id.icon_frame).setBackground(background);

        ((AppCompatTextView) rootView.findViewById(R.id.title)).setText(name);

        SwitchCompat sc = rootView.findViewById(R.id.cb);
        sc.setText(getString(R.string.display_analysis_tag_status, typeName.toLowerCase()));
        sc.setChecked(prefs.getBoolean(type, true));
        sc.setOnCheckedChangeListener((buttonView, isChecked) -> prefs.edit().putBoolean(type, isChecked).apply());

        Spanned messageToBeShown = Html.fromHtml(getString(R.string.ingredients_in_this_product_are, name.toLowerCase()));
        AppCompatButton helpNeeded = rootView.findViewById(R.id.helpNeeded);
        boolean showHelpTranslate = tag != null && tag.contains("unknown");
        AppCompatImageView image = rootView.findViewById(R.id.image);
        if (arguments.getBoolean(PHOTOS_TO_BE_VALIDATED_KEY, false)) {
            messageToBeShown = Html.fromHtml(getString(R.string.unknown_status_missing_ingredients));
            image.setImageResource(R.drawable.ic_add_a_photo_dark_48dp);
            image.setOnClickListener(v -> goToAddPhoto());
            helpNeeded.setText(Html.fromHtml(getString(R.string.add_photo_to_extract_ingredients)));
            helpNeeded.setOnClickListener(v -> goToAddPhoto());
        } else if (tag != null && ambiguousIngredient != null) {
            messageToBeShown = Html.fromHtml(getString(R.string.unknown_status_ambiguous_ingredients, ambiguousIngredient));
            helpNeeded.setVisibility(View.GONE);
        } else if (showHelpTranslate && arguments.getBoolean(MISSING_INGREDIENTS_KEY, false)) {
            String ingredientsImageUrl = arguments.getString(INGREDIENTS_IMAGE_URL_KEY);
            Utils.picassoBuilder(getActivity())
                .load(ingredientsImageUrl)
                .into(image);
            image.setOnClickListener(v -> goToExtract());
            messageToBeShown = Html.fromHtml(getString(R.string.unknown_status_missing_ingredients));
            helpNeeded.setText(Html.fromHtml(getString(R.string.help_extract_ingredients, typeName.toLowerCase())));
            helpNeeded.setOnClickListener(v -> goToExtract());
            helpNeeded.setVisibility(View.VISIBLE);
        } else if (showHelpTranslate) {
            messageToBeShown = Html.fromHtml(getString(R.string.unknown_status_no_translation));
            helpNeeded.setText(Html.fromHtml(getString(R.string.help_translate_ingredients)));
            helpNeeded.setOnClickListener(v -> {
                CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder().build();
                CustomTabActivityHelper.openCustomTab(
                    getActivity(),// activity
                    customTabsIntent,
                    Uri.parse(getString(R.string.help_translate_ingredients_link, Locale.getDefault().getLanguage())),
                    (activity, uri) -> {
                        Intent i = new Intent(Intent.ACTION_VIEW);
                        i.setData(uri);
                        activity.startActivity(i);
                    }
                );
            });
            helpNeeded.setVisibility(View.VISIBLE);
        } else {
            image.setVisibility(View.GONE);
            if (!TextUtils.isEmpty(ingredients)) {
                messageToBeShown = Html.fromHtml(getString(R.string.ingredients_in_this_product, name.toLowerCase()) + ingredients);
            }
            helpNeeded.setVisibility(View.GONE);
        }
        AppCompatTextView message = rootView.findViewById(R.id.message);
        message.setText(messageToBeShown);

        rootView.findViewById(R.id.close).setOnClickListener(v -> dismiss());
        return rootView;
    }

    private void goToAddPhoto() {
        dismiss();
        if (getActivity() instanceof ContinuousScanActivity) {
            ((ContinuousScanActivity) getActivity()).showIngredientsTab(ProductViewActivity.ShowIngredientsAction.SEND_UPDATED);
        } else if (getActivity() instanceof ProductViewActivity) {
            ((ProductViewActivity) getActivity()).showIngredientsTab(ProductViewActivity.ShowIngredientsAction.SEND_UPDATED);
        }
    }

    private void goToExtract() {
        dismiss();
        if (getActivity() instanceof ContinuousScanActivity) {
            ((ContinuousScanActivity) getActivity()).showIngredientsTab(ProductViewActivity.ShowIngredientsAction.PERFORM_OCR);
        } else if (getActivity() instanceof ProductViewActivity) {
            ((ProductViewActivity) getActivity()).showIngredientsTab(ProductViewActivity.ShowIngredientsAction.PERFORM_OCR);
        }
    }

    public void setOnDismissListener(DialogInterface.OnDismissListener onDismissListener) {
        this.onDismissListener = onDismissListener;
    }

    @Override
    public void onDismiss(@NonNull DialogInterface dialog) {
        super.onDismiss(dialog);

        if (onDismissListener != null) {
            onDismissListener.onDismiss(dialog);
        }
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/features/product/view/ingredients_analysis/IngredientsWithTagDialogFragment.kt
package openfoodfacts.github.scrachx.openfood.features.product.view.ingredients_analysis

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.Color
import android.net.Uri
import android.os.Bundle
import android.text.TextUtils
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.AppCompatTextView
import androidx.browser.customtabs.CustomTabsIntent
import androidx.core.content.edit
import androidx.core.content.res.ResourcesCompat
import androidx.core.graphics.BlendModeColorFilterCompat
import androidx.core.graphics.BlendModeCompat
import androidx.core.text.HtmlCompat
import androidx.fragment.app.DialogFragment
import com.squareup.picasso.Picasso
import dagger.hilt.android.AndroidEntryPoint
import openfoodfacts.github.scrachx.openfood.R
import openfoodfacts.github.scrachx.openfood.customtabs.CustomTabActivityHelper
import openfoodfacts.github.scrachx.openfood.databinding.IngredientsWithTagBinding
import openfoodfacts.github.scrachx.openfood.features.product.view.IProductView
import openfoodfacts.github.scrachx.openfood.features.product.view.ProductViewActivity
import openfoodfacts.github.scrachx.openfood.models.Product
import openfoodfacts.github.scrachx.openfood.models.entities.analysistagconfig.AnalysisTagConfig
import java.util.*
import javax.inject.Inject

@AndroidEntryPoint
class IngredientsWithTagDialogFragment : DialogFragment() {
    private var _binding: IngredientsWithTagBinding? = null
    private val binding get() = _binding!!

    @Inject
    lateinit var picasso: Picasso

    @Inject
    lateinit var sharedPreferences: SharedPreferences

    var onDismissListener: ((DialogInterface) -> Unit)? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = IngredientsWithTagBinding.inflate(inflater, container, false)
        requireDialog().window!!.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        requireDialog().window!!.setGravity(Gravity.CENTER)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val arguments = requireArguments()
        val tag = arguments.getString(TAG_KEY)
        val type = arguments.getString(TYPE_KEY)
        val typeName = arguments.getString(TYPE_NAME_KEY) ?: "UNKNOWN"
        val iconUrl = arguments.getString(ICON_URL_KEY)
        val color = arguments.getString(COLOR_KEY)
        val name = arguments.getString(NAME_KEY)
        val ingredients = arguments.getString(INGREDIENTS_KEY)
        val ambiguousIngredient = arguments.getString(AMBIGUOUS_INGREDIENT_KEY)
        val ingredientsImageUrl = arguments.getString(INGREDIENTS_IMAGE_URL_KEY)

        picasso
            .load(iconUrl)
            .into(binding.icon)
        binding.iconFrame.background =
            ResourcesCompat.getDrawable(requireActivity().resources, R.drawable.rounded_button, requireActivity().theme)
                ?.apply {
                    colorFilter = BlendModeColorFilterCompat.createBlendModeColorFilterCompat(
                        Color.parseColor(color),
                        BlendModeCompat.SRC_IN
                    )
                }
        binding.title.text = name
        binding.cb.let {
            it.text = getString(R.string.display_analysis_tag_status, typeName.lowercase(Locale.getDefault()))
            it.isChecked = sharedPreferences.getBoolean(type, true)
            it.setOnCheckedChangeListener { _, isChecked ->
                sharedPreferences.edit { putBoolean(type, isChecked) }
            }
        }
        var messageToBeShown =
            HtmlCompat.fromHtml(
                getString(
                    R.string.ingredients_in_this_product_are,
                    name!!.lowercase(Locale.getDefault())
                ), HtmlCompat.FROM_HTML_MODE_LEGACY
            )
        val showHelpTranslate = tag != null && tag.contains("unknown")

        if (arguments.getBoolean(PHOTOS_TO_BE_VALIDATED_KEY, false)) {
            messageToBeShown = HtmlCompat.fromHtml(
                getString(R.string.unknown_status_missing_ingredients),
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )
            binding.image.setImageResource(R.drawable.ic_add_a_photo_dark_48dp)
            binding.image.setOnClickListener { goToAddPhoto() }
            binding.helpNeeded.text = HtmlCompat.fromHtml(
                getString(R.string.add_photo_to_extract_ingredients),
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )
            binding.helpNeeded.setOnClickListener { goToAddPhoto() }
        } else if (tag != null && ambiguousIngredient != null) {
            messageToBeShown =
                HtmlCompat.fromHtml(
                    getString(R.string.unknown_status_ambiguous_ingredients, ambiguousIngredient),
                    HtmlCompat.FROM_HTML_MODE_LEGACY
                )
            binding.helpNeeded.visibility = View.GONE
        } else if (showHelpTranslate && arguments.getBoolean(MISSING_INGREDIENTS_KEY, false)) {
            picasso.load(ingredientsImageUrl).into(binding.image)

            binding.image.setOnClickListener { goToExtract() }
            messageToBeShown = HtmlCompat.fromHtml(
                getString(R.string.unknown_status_missing_ingredients),
                HtmlCompat.FROM_HTML_MODE_LEGACY
            )

            binding.helpNeeded.text =
                HtmlCompat.fromHtml(
                    getString(
                        R.string.help_extract_ingredients,
                        typeName.lowercase(Locale.getDefault())
                    ), HtmlCompat.FROM_HTML_MODE_LEGACY
                )
            binding.helpNeeded.setOnClickListener { goToExtract() }
            binding.helpNeeded.visibility = View.VISIBLE

        } else if (showHelpTranslate) {
            messageToBeShown =
                HtmlCompat.fromHtml(getString(R.string.unknown_status_no_translation), HtmlCompat.FROM_HTML_MODE_LEGACY)
            binding.helpNeeded.text =
                HtmlCompat.fromHtml(getString(R.string.help_translate_ingredients), HtmlCompat.FROM_HTML_MODE_LEGACY)
            binding.helpNeeded.setOnClickListener {
                val customTabsIntent = CustomTabsIntent.Builder().build()
                CustomTabActivityHelper.openCustomTab(
                    requireActivity(),  // activity
                    customTabsIntent,
                    Uri.parse(getString(R.string.help_translate_ingredients_link, Locale.getDefault().language))
                ) { activity: Activity, uri: Uri ->
                    val i = Intent(Intent.ACTION_VIEW)
                    i.data = uri
                    activity.startActivity(i)
                }
            }
            binding.helpNeeded.visibility = View.VISIBLE
        } else {
            binding.image.visibility = View.GONE
            if (!TextUtils.isEmpty(ingredients)) {
                messageToBeShown = HtmlCompat.fromHtml(
                    "${
                        getString(
                            R.string.ingredients_in_this_product,
                            name.lowercase(Locale.getDefault())
                        )
                    }$ingredients",
                    HtmlCompat.FROM_HTML_MODE_LEGACY
                )
            }
            binding.helpNeeded.visibility = View.GONE
        }
        val message: AppCompatTextView = binding.message
        message.text = messageToBeShown
        binding.close.setOnClickListener { dismiss() }
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    private fun goToAddPhoto() {
        dismiss()
        (activity as? IProductView)?.showIngredientsTab(ProductViewActivity.ShowIngredientsAction.SEND_UPDATED)
    }

    private fun goToExtract() {
        dismiss()
        (activity as? IProductView)?.showIngredientsTab(ProductViewActivity.ShowIngredientsAction.PERFORM_OCR)
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        onDismissListener?.invoke(dialog)
    }

    companion object {
        private const val TAG_KEY = "tag"
        private const val TYPE_KEY = "type"
        private const val TYPE_NAME_KEY = "type_name"
        private const val ICON_URL_KEY = "icon_url"
        private const val COLOR_KEY = "color"
        private const val NAME_KEY = "name"
        private const val INGREDIENTS_IMAGE_URL_KEY = "ingredients_image_url"
        private const val INGREDIENTS_KEY = "ingredients"
        private const val PACKAGING_IMAGE_URL_KEY = "packaging_image_url"
        private const val MISSING_INGREDIENTS_KEY = "missing_ingredients"
        private const val PHOTOS_TO_BE_VALIDATED_KEY = "photos_to_be_validated"
        private const val AMBIGUOUS_INGREDIENT_KEY = "ambiguous_ingredient"

        @JvmStatic
        fun newInstance(product: Product, config: AnalysisTagConfig) = IngredientsWithTagDialogFragment().apply {
            arguments = Bundle().apply {
                putString(TAG_KEY, config.analysisTag)
                putString(TYPE_KEY, config.type)
                putString(TYPE_NAME_KEY, config.typeName)
                putString(ICON_URL_KEY, config.iconUrl)
                putString(COLOR_KEY, config.color)
                putString(NAME_KEY, config.name.name)
                putString(INGREDIENTS_IMAGE_URL_KEY, product.imageIngredientsUrl)

                if (product.ingredients.isEmpty()) {
                    val statesTags = product.statesTags
                    var ingredientsToBeCompleted = false
                    var photosToBeValidated = false
                    statesTags.forEach { stateTag ->
                        ingredientsToBeCompleted = stateTag == "en:ingredients-to-be-completed"
                        photosToBeValidated = stateTag == "en:photos-to-be-validated"
                    }
                    if (ingredientsToBeCompleted && photosToBeValidated) {
                        putBoolean(PHOTOS_TO_BE_VALIDATED_KEY, true)
                    } else {
                        putBoolean(MISSING_INGREDIENTS_KEY, true)
                    }
                } else {
                    val showIngredients = config.name.showIngredients
                    if (showIngredients != null) {
                        putSerializable(
                            INGREDIENTS_KEY,
                            getMatchingIngredientsText(product, showIngredients.split(":").toTypedArray())
                        )
                    }
                    val ambiguousIngredient = product.ingredients
                        .filter {
                            it.additionalProperties.containsKey(config.type) && it.additionalProperties.containsValue(
                                "maybe"
                            )
                        }
                        .mapNotNull { it.text }


                    if (ambiguousIngredient.isNotEmpty()) {
                        putString(AMBIGUOUS_INGREDIENT_KEY, ambiguousIngredient.joinToString(","))
                    }
                }
            }
        }

        private fun getMatchingIngredientsText(
            product: Product,
            ingredients: Array<String>,
        ): String? {

            val matchingIngredients = product.ingredients
                .filter { ingredients[1] == it.additionalProperties[ingredients[0]] }
                .mapNotNull { it.text }
                .map { it.lowercase(Locale.getDefault()).replace("_", "") }


            return if (matchingIngredients.isEmpty()) null
            else StringBuilder().apply {
                append(" <b>")
                matchingIngredients.withIndex().forEach { (index, ingredient) ->
                    if (index != 0) append(", ")
                    append(ingredient)
                }
                append("</b>")
            }.toString()
        }
    }
}
